### PR TITLE
Feature: Get user by email endpoint

### PR DESCRIPTION
**Pull request by Miles-Alford**
### *What is the problem we're trying to solve*
This PR adds the endpoint to allow getting a user by their email. The purpose of this endpoint is to allow the front-end to get the logged-in user when viewing projects they are assigned on as the front-end has no way of knowing the user ID until this endpoint is called to return all the user information to the front-end.

### *What changes have we introduced*

Added the complete endpoint and its respective tests (controller, usecase, gateway, request). Also validation has been added to the POST endpoint to now first check if a user exists before creating a new one.

### *Follow up actions after merging PR*

No follow up needed

### CODE REVIEW

File: ProjectFinderApi.Tests/V1/UseCase/UsersUseCaseTests.cs [hunk 1]
```diff
@@ -58,6 +60,18 @@ public void ExecutePostReturnsCreatedUser()
...
+        [Test]
+        public void ExecutePostThrowsExceptionIfUserAlreadyExists()
+        {
+            var userRequest = TestHelpers.CreateUserRequest();
+            var user = TestHelpers.CreateUser(email: userRequest.EmailAddress);
+            _mockUsersGateway.Setup(x => x.GetUserByEmail(userRequest.EmailAddress)).Returns(user);
+
+            Action act = () => _usersUseCase.ExecutePost(userRequest);
+
+            act.Should().Throw<PostUserException>().WithMessage($"User with email {userRequest.EmailAddress} already exists on thr system");
```
Comment (LBHTKarki) [inline ProjectFinderApi.Tests/V1/UseCase/UsersUseCaseTests.cs:72]: small typo 'thr'

File: ProjectFinderApi/V1/Controllers/UserController.cs [hunk 2]
```diff
@@ -65,5 +65,32 @@ public IActionResult GetUsers()
 
             return Ok(users);
         }
+
+        /// <summary>
+        /// Get an individual user by their email
+        /// </summary>
+        /// <response code="400">Invalid GetUserRequest received.</response>
+        /// <response code="404">No user with that email found.</response>
+        [ProducesResponseType(typeof(UserResponse), StatusCodes.Status200OK)]
```
Comment (LBHTKarki) [inline ProjectFinderApi/V1/Controllers/UserController.cs:74]: perhaps include 200 response as well

File: ProjectFinderApi/V1/Controllers/UserController.cs [hunk 3]
```diff
@@ -65,5 +65,32 @@ public IActionResult GetUsers()
...
+        [HttpGet]
+        [Route("email")]
+        public IActionResult GetUserByEmail([FromBody] GetUserRequest request)
```
Comment (LBHTKarki) [inline ProjectFinderApi/V1/Controllers/UserController.cs:77]: should this be [FromQuery] ?

File: ProjectFinderApi/V1/UseCase/UsersUseCase.cs [hunk 4]
```diff
@@ -21,6 +22,11 @@ public UsersUseCase(IUsersGateway usersGateway)
         public UserResponse ExecutePost(CreateUserRequest createUserRequest)
         {
             // Check if user exists first, return error if so
+            var existingUser = _usersGateway.GetUserByEmail(createUserRequest.EmailAddress);
+            if (existingUser != null)
+            {
+                throw new PostUserException($"User with email {createUserRequest.EmailAddress} already exists on thr system");
```
Comment (LBHTKarki) [inline ProjectFinderApi/V1/UseCase/UsersUseCase.cs:28]: small typo 'thr'

File: ProjectFinderApi/V1/Gateways/UsersGateway.cs [hunk 5]
```diff
@@ -35,5 +36,10 @@ public IEnumerable<User> GetUsers()
         {
             return _databaseContext.Users;
         }
+
+        public User GetUserByEmail(string email)
+        {
+            return _databaseContext.Users.Where(user => user.Email.ToLower() == email.ToLower()).FirstOrDefault();
```
Comment (LBHTKarki) [inline ProjectFinderApi/V1/Gateways/UsersGateway.cs:42]: Can user.Email ever be null here? If not, then this is fine

Comment (Miles-Alford) [reply ProjectFinderApi/V1/Gateways/UsersGateway.cs:42]: `user.Email` should never be null. All users must have a email to successfully insert into the database

File: ProjectFinderApi/V1/Controllers/UserController.cs (hunk 2 above)
Comment (Miles-Alford) [reply ProjectFinderApi/V1/Controllers/UserController.cs:75]: 👍

**Review by LBHTKarki — APPROVED** [review]
Looks good 👍

**Pull request merged**